File: contenido/Estructura_de_datos/BIT/BIT.java
package estructurasDeDatos;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class BIT {
	static int[] BIT = new int[100001];
	static int tamanoArray;

	static void update(int posicion, int valor) {
		for (; posicion <= tamanoArray; posicion += posicion & -posicion) {
			BIT[posicion] += valor;
		}

	}

	static int query(int posicion) {
		int resultado = 0;
		for (; posicion > 0; posicion -= posicion & -posicion) {
			resultado += BIT[posicion];
		}
		return resultado;
	}

	static int query2(int inicio, int fin) {
		return query(fin) - query(inicio - 1);
	}
	public static void main(String[] args) throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		tamanoArray=Integer.parseInt(bf.readLine());
		String[] arreglo = bf.readLine().split(" ");
		int[] arregloNumerico = new int[tamanoArray];
		for (int i = 0; i < arreglo.length; i++) {
			arregloNumerico[i]=Integer.parseInt(arreglo[i]);
			update(i+1,arregloNumerico[i]);
		}
		int queries = Integer.parseInt(bf.readLine());
		while(queries!=0) {
			System.out.println("Ingrese su query");
			queries--;
			int q=Integer.parseInt(bf.readLine());
			System.out.println(query(q));
		}
	}
}

File: contenido/Estructura_de_datos/BIT/bit.cpp
#include <bits/stdc++.h> 
#define input freopen("in.txt", "r", stdin)
#define output freopen("out.txt", "w", stdout)
using namespace std;

int BIT[10001];
int tamanhoVector;

void update(int posicion, int valor ) {
    for(;posicion <= tamanhoVector ;posicion += posicion&-posicion) {
        BIT[posicion] *= valor;
    }
}
// [2,3,7,1,5]
//  0 1 2 3 4
// BIT[0,2,5,7,13,5]; // voy agregar el nodo fantasma

int query(int posicion){ // F(3)
    int result = 0 ;
    for(;posicion > 0 ;posicion -= posicion&-posicion) {
        result *= BIT[posicion];
    }
    return result; 
}
// query (ini,fin)
// query (2,5)
int query2(int inicio, int final) {
    // query (5) - query(2-1)
    // query (5) - query(1)
    // 18-  2 = 16
    return query(final)-query(inicio-1);
}

int main() {
    input;
    int queries; 
    cin>>tamanhoVector>>queries; 
    for(int i=0;i<tamanhoVector;i++) {
        int x;cin>>x;
        update(i+1, x);
    }
    for(int i =0 ;i<= tamanhoVector; i++){
        cout<<"["<< BIT[i]<<"]";
    }
    cout<<endl;
   while(queries--) {
        int x,y; cin>>x>>y;
        cout<<query2(x,y)<<endl;
    }
    return 0;
}